xarray/tests/test_variable.py

```diff
@@ -2140,6 +2140,15 @@ def assert_assigned_2d(array, key_x, key_y, values):
         expected = Variable(["x", "y"], [[2, 3], [3, 4], [4, 5]])
         assert_identical(v, expected)
 
+    def test_setitem_with_dataarray(self):
+        v = Variable(["a", "b", "c", "d"], np.r_[:120].reshape(2, 3, 4, 5))
+        b = Variable(["u", "v"], [[0, 0], [1, 0]])
+        c = Variable(["u", "v"], [[0, 1], [2, 3]])
+        w = DataArray([-1, -2], dims=["u"])
+        index = dict(b=b, c=c)
+        v[index] = w
```

Comment:
I kind of assumed we want to be able to do `variable[...] = dataarray`. If not, we could instead put the fix in `DataArray.__setitem__`.

Reply:
> we could instead put the fix in DataArray.__setitem__.

Yes I think this is the right way and would fit your intuition (previous comment) of the fix being a bit ugly

Reply:
Ok, got it! Should be addressed now.